client/client.go

```diff
@@ -1,71 +1,378 @@
+				return err
+			}
+		}
+		if err := signed.Verify(s, "root", 0, db); err != nil {
+			return err
+		}
+		c.root = root
+		c.db = db
+	} else {
+		return ErrNoRootKeys
+	}
 
-	/*
-		If at any point in the following process there is a problem (e.g., only expired
-		metadata can be retrieved), the Root file is downloaded and the process starts
-		over. Optionally, the software update system using the framework can decide how
-		to proceed rather than automatically downloading a new Root file.
+	if snapshotJSON, ok := meta["snapshot.json"]; ok {
+		snapshot := &data.Snapshot{}
+		if err := signed.Unmarshal(snapshotJSON, snapshot, "snapshot", 0, c.db); err != nil {
+			return err
+		}
+		c.snapshot = snapshot
+	}
 
-			TUF downloads and verifies timestamp.json.
+	if targetsJSON, ok := meta["targets.json"]; ok {
+		targets := &data.Targets{}
+		if err := signed.Unmarshal(targetsJSON, targets, "targets", 0, c.db); err != nil {
+			return err
+		}
+		c.targets = targets
+	}
 
-			If timestamp.json indicates that snapshot.json has changed, TUF downloads
-			and verifies snapshot.json.
+	if timestampJSON, ok := meta["timestamp.json"]; ok {
+		timestamp := &data.Timestamp{}
+		if err := signed.Unmarshal(timestampJSON, timestamp, "timestamp", 0, c.db); err != nil {
+			return err
+		}
+		c.timestamp = timestamp
+	}
 
-			TUF determines which metadata files listed in snapshot.json differ from
-			those described in the last snapshot.json that TUF has seen. If root.json
-			has changed, the update process starts over using the new root.json.
+	c.localMeta = meta
+	return nil
+}
 
-			TUF provides the software update system with a list of available files
-			according to targets.json.
-	*/
+// downloadMeta downloads top-level metadata from remote storage and verifies
+// it using the given file metadata.
+func (c *Client) downloadMeta(name string, m *data.FileMeta) ([]byte, error) {
+	r, err := c.remote.Get(name)
+	if err != nil {
+		if err == ErrNotFound {
+			return nil, ErrMissingRemoteMetadata{name}
+		}
+		return nil, err
+	}
+	defer r.Close()
+
+	// if m is nil (e.g. when downloading timestamp.json, which has unknown
+	// size), just read the entire stream and return it
+	if m == nil {
+		b, err := ioutil.ReadAll(r)
+		if err != nil {
+			return nil, err
+		}
+		return b, nil
+	}
+
+	// read exactly m.Length bytes from the stream
+	buf := make([]byte, m.Length)
+	if _, err := io.ReadFull(r, buf); err != nil {
+		return nil, err
+	}
+
+	// verify buf matches given metadata
+	meta, err := util.GenerateFileMeta(bytes.NewReader(buf))
+	if err != nil {
+		return nil, err
+	}
+	if !util.FileMetaEqual(meta, *m) {
+		return nil, ErrChecksumFailed
+	}
+	return buf, nil
+}
 
+// decodeRoot decodes and verifies root metadata.
+func (c *Client) decodeRoot(b json.RawMessage) error {
+	var minVer int
+	if c.root != nil {
+		minVer = c.root.Version
+	}
+	if err := signed.Unmarshal(b, &data.Root{}, "root", minVer, c.db); err != nil {
```

Comment:
What is this doing, it doesn't seem to save the data? Perhaps it should be called `verifyRoot` or similar? Also you can just `return` the result of the unmarshal.